design/high-level-reactivity.md

```diff
@@ -0,0 +1,245 @@
+        }
+      }
+    ]
+  }
+}
+```
+
+This will tell the client which dependencies they should watch to figure out if the list object itself changed, or if the list of tasks needs to be updated.
+
+Note that can’t put the `_deps` field on tasks itself, because JSON doesn’t allow it, so we have to put it on the parent instead. Also, the `_self` field is a shorthand to avoid listing all of the properties of the list object - it could have `name`, `description`, and so on, and it would be a waste of bandwidth to send all of those key names again.
+
+### Automatically recording dependencies when reading data
+In order to know when a GraphQL query needs to be re-run, we need to know what deps represent different parts of the query. These deps can be recorded manually for complex queries, but simple queries can have their deps identified automatically. For example, here’s a query used in a particular part of the GraphQL resolution tree:
+
+```
+todoLists.select('*').where('id', 1);
+```
+
+This might record the following dependency:
+
+```
+{ key: 'todoLists:1', version: 0 }
+```
+
+The dependency recording mechanism queries the dependency tracking server to figure out the current version.
+
+### Manually recording dependencies
+
+The automatic dependency recording mechanism won’t work if it isn’t possible to analyze the query to determine the dependencies. In these cases, the developer will need to manually record a dependency, using any string key they want.
+
+For example, if you have a complex query to count the number of notifications a user has, you might want a custom invalidation key for that number:
+
+```
+// somewhere, a function that consistently generates a key
+function notificationDepKeyForUser(userId) {
+  return 'notificationCount:' + userId;
+}
+
+// inside the GraphQL resolver
+numNotifications = getNotificationCountForUser(userId);
+context.recordDependency(notificationDepKeyForUser(userId));
+```
+
+This will let you manually specify situations in which the number of notifications should be refreshed. Advanced users with particularly strict performance requirements, like a site-wide visitor count or a realtime high-score table, might also choose to use manually constructed dependencies for more control over their data flow.
+
+## Simple reactivity model
+
+Based on the above primitives, here is a stateless strategy for GraphQL reactivity:
+
+1. The client fetches the query from the GraphQL server, which includes a set of deps in the response.
+2. The client periodically polls the invalidation server with the set of deps, and the server returns the list of deps which have a newer version available. Note that this can easily be converted to a stateful approach where the client subscribes to a list of depenencies over a websocket, for parts of the client which need lower latency - see the [section below](#reducing-latency).
+3. The client re-fetches the subtrees of the query which depend on the invalidated deps.
```

Comment:
Meteor currently puts in a lot of effort on the server to diff data and try very hard to avoid sending any extra bytes. We think in many cases it might be better to send a little bit of extra data if that saves a ton of CPU and RAM.

So it's true that this system might send too much new data, but you can optimize specific situations by controlling where the dependency keys show up.

Once we have more details, let's talk to make sure this is aligned with what you need. I think it will be, but we should make sure!